From a79a195e05e9fce36bdf7db5923da2b5c88bd3a3 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:13 -0400
Subject: [ethernet] qlcnic: Update ethtool standard pause settings

Message-id: <1408626422-3941-30-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88570
O-Subject: [PATCH RHEL 7.1 029/138] qlcnic: Update ethtool standard pause settings.
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 6177a95a93fe6eed2f59fa17720057988a81913c

Update ethtool standard pause parameter settings and display

Signed-off-by: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>
Signed-off-by: Himanshu Madhani <himanshu.madhani@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
index 56c1572..37f3c47 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
@@ -3369,10 +3369,21 @@ void qlcnic_83xx_get_pauseparam(struct qlcnic_adapter *adapter,
  }
  config = ahw->port_config;
  if (config & QLC_83XX_CFG_STD_PAUSE) {
-  if (config & QLC_83XX_CFG_STD_TX_PAUSE)
+  switch (MSW(config)) {
+  case QLC_83XX_TX_PAUSE:
+   pause->tx_pause = 1;
+   break;
+  case QLC_83XX_RX_PAUSE:
+   pause->rx_pause = 1;
+   break;
+  case QLC_83XX_TX_RX_PAUSE:
+  default:
+   /* Backward compatibility for existing
+    * flash definitions
+    */
    pause->tx_pause = 1;
-  if (config & QLC_83XX_CFG_STD_RX_PAUSE)
    pause->rx_pause = 1;
+  }
  }
 
  if (QLC_83XX_AUTONEG(config))
@@ -3415,7 +3426,8 @@ int qlcnic_83xx_set_pauseparam(struct qlcnic_adapter *adapter,
   ahw->port_config &= ~QLC_83XX_CFG_STD_RX_PAUSE;
   ahw->port_config |= QLC_83XX_CFG_STD_TX_PAUSE;
  } else if (!pause->rx_pause && !pause->tx_pause) {
-  ahw->port_config &= ~QLC_83XX_CFG_STD_TX_RX_PAUSE;
+  ahw->port_config &= ~(QLC_83XX_CFG_STD_TX_RX_PAUSE |
+          QLC_83XX_CFG_STD_PAUSE);
  }
  status = qlcnic_83xx_set_port_config(adapter);
  if (status) {
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.h b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.h
index 533e150..2883b57 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.h
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.h
@@ -363,6 +363,9 @@ enum qlcnic_83xx_states {
 #define QLC_83XX_LINK_EEE(data)  ((data) & BIT_13)
 #define QLC_83XX_DCBX(data)   (((data) >> 28) & 7)
 #define QLC_83XX_AUTONEG(data)   ((data) & BIT_15)
+#define QLC_83XX_TX_PAUSE   0x10
+#define QLC_83XX_RX_PAUSE   0x20
+#define QLC_83XX_TX_RX_PAUSE   0x30
 #define QLC_83XX_CFG_STD_PAUSE   (1 << 5)
 #define QLC_83XX_CFG_STD_TX_PAUSE  (1 << 20)
 #define QLC_83XX_CFG_STD_RX_PAUSE  (2 << 20)
-- 
1.7.1